AmplifyPlugins/Core/AWSPluginsCore/Model/Support/GraphQLDocumentnputValue.swift

```diff
@@ -10,19 +10,34 @@ import Foundation
 /// A container to hold either an object or a value, useful for storing document inputs and allowing manipulation at
 /// the first level of the object
 public enum GraphQLDocumentInputValue {
-    case scalarOrString(Any)
+    case scalar(GraphQLDocumentValueRepresentable)
     case object([String: Any?])
 }
 
-/// Contains the `type` of the GraphQL document input parameter as a string value and `GraphQLDocumentInputValue`
-public struct GraphQLDocumentInput {
+public protocol GraphQLDocumentValueRepresentable {
+    var graphQLDocumentValue: String { get }
+}
+
+extension Int: GraphQLDocumentValueRepresentable {
+    public var graphQLDocumentValue: String {
+        return "\(self)"
+    }
+}
 
-    public var type: String
+extension String: GraphQLDocumentValueRepresentable {
+    public var graphQLDocumentValue: String {
+        return self
+    }
+}
 
-    public var value: GraphQLDocumentInputValue
+extension Bool: GraphQLDocumentValueRepresentable {
+    public var graphQLDocumentValue: String {
+        return "\(self)"
+    }
+}
 
-    public init(type: String, value: GraphQLDocumentInputValue) {
-        self.type = type
-        self.value = value
+extension Decimal: GraphQLDocumentValueRepresentable {
```

Comment:
I'm surprised we're using Decimal--I thought we had moved over to Float for our codegenned models?

Reply:
I believe we landed on Double because it can hold up to 64 bit. then i encountered some serialization issues, a Double with value 12.3 is serialized as "12.300000000001", which is why i started using Decimal (https://github.com/aws-amplify/amplify-ios/issues/159). so still not 100% confident in which is the right type to use as looking up more details on Decimal isn't giving me many results to resolve this. Maybe there to use Double and override the serialization logic on Doubles. I think the best test is to write an integration test which has a single model with all of the various types that we support.

I believe Float in AppSync is currently returning .double as the ModelFieldType, which then gets translated to a Decimal and serialized. https://github.com/aws-amplify/amplify-ios/blob/461c3cfd8923fbbb6eb10b60495b0f4594879cf4/Amplify/Categories/DataStore/Model/ModelSchema%2BDefinition.swift#L56

Reply:
opening an issue track adding test with all types that are supported, may be completed as part of @drochetti 's work or follow up task https://github.com/aws-amplify/amplify-ios/issues/314